plugins/animate/squeezimize.hpp

```diff
@@ -0,0 +1,388 @@
+        {
+            output = view->get_output();
+            output->render->add_effect(&pre_hook, wf::OUTPUT_EFFECT_PRE);
+        }
+
+        animation_geometry.x     = std::min(bbox.x, minimize_target.x);
+        animation_geometry.y     = std::min(bbox.y, minimize_target.y);
+        animation_geometry.width =
+            std::max(std::max(std::max(bbox.width,
+                minimize_target.width),
+                (minimize_target.x + minimize_target.width) - bbox.x),
+                (bbox.x + bbox.width) - minimize_target.x);
+        animation_geometry.height =
+            std::max(std::max(std::max(bbox.height,
+                minimize_target.height),
+                (minimize_target.y + minimize_target.height) - bbox.y),
+                (bbox.y + bbox.height) - minimize_target.y);
+        OpenGL::render_begin();
+        program.compile(squeeze_vert_source, squeeze_frag_source);
+        OpenGL::render_end();
+    }
+
+    wf::geometry_t get_bounding_box() override
+    {
+        return this->animation_geometry;
+    }
+
+    wf::effect_hook_t pre_hook = [=] ()
+    {
+        output->render->damage(animation_geometry);
+    };
+
+    void gen_render_instances(std::vector<render_instance_uptr>& instances,
+        damage_callback push_damage, wf::output_t *shown_on) override
+    {
+        instances.push_back(std::make_unique<simple_node_render_instance_t>(
+            this, push_damage, view));
+    }
+
+    void init_animation(bool squeeze)
+    {
+        if (!squeeze)
+        {
+            this->progression.reverse();
+        }
+
+        this->progression.start();
+    }
+
+    virtual ~squeezimize_transformer()
+    {
+        if (output)
+        {
+            output->render->rem_effect(&pre_hook);
+        }
+
+        program.free_resources();
+    }
+};
+}
+}
+
+class squeezimize_animation : public animation_base
+{
+    wayfire_view view;
+
+  public:
+    void init(wayfire_view view, wf::animation_description_t dur, wf_animation_type type) override
+    {
+        this->view = view;
+        pop_transformer(view);
+        auto bbox     = view->get_transformed_node()->get_children_bounding_box();
+        auto toplevel = wf::toplevel_cast(view);
+        wf::dassert(toplevel != nullptr, "We cannot minimize non-toplevel views!");
+        auto hint = toplevel->get_minimize_hint();
+        auto tmgr = view->get_transformed_node();
+        auto node = std::make_shared<wf::squeezimize::squeezimize_transformer>(view, hint, bbox);
+        tmgr->add_transformer(node, wf::TRANSFORMER_HIGHLEVEL + 1, sqeezimize_transformer_name);
+        node->init_animation(type & HIDING_ANIMATION);
+    }
+
+    void pop_transformer(wayfire_view view)
+    {
+        if (view->get_transformed_node()->get_transformer(sqeezimize_transformer_name))
+        {
+            view->get_transformed_node()->rem_transformer(sqeezimize_transformer_name);
+        }
+    }
+
+    bool step() override
+    {
+        if (!view)
```

Comment:
Is it possible that view becomes null? It seems to me that this shouldn't be necessary.

Reply:
Probably not necessary.